docs/design/sum_types.md

```diff
@@ -0,0 +1,197 @@
+
+```carbon
+choice OptionalI32 {
+  Some(value: i32),
+  None
+}
+```
+
+This declares a sum type named `OptionalI32` with two alternatives: `Some`,
+which holds a single `i32` value, and `None`, which is empty. Choice types can
+also be parameterized, like class types:
+
+```carbon
+choice Optional(T:! Type) {
+  Some(value: T),
+  None
+}
+```
+
+An alternative declaration consists of the alternative name, followed by an
+optional parameter list in parentheses, much like the syntax of a function
+declaration. A value of a function-like alternative is specified by "calling" it
+like a function, and a value of a "bare" alternative like `None` is specified by
+naming it:
+
+```carbon
+var my_opt: Optional(i32) = Optional(i32).None;
+my_opt = Optional(i32).Some(42);
+```
+
+The value of a choice type can be inspected using a `match` statement:
+
+```carbon
+match (my_opt) {
+  case .Some(the_value: i32) => {
+    Print(the_value);
+  }
+  case .None => {
+    Print("None");
+  }
+}
+```
+
+## User-defined sum types
+
+`choice` declarations are a convenience shorthand for common use cases, but they
+have limited flexibility. There is no way to control the representation of a
+`choice` type, or define methods or other members for it (although you can
+extend it to implement interfaces, using
+[`external impl`](generics/overview.md#implementing-interfaces) or
+[`adapter`](generics/overview.md#adapting-types)). However, a `class` type can
+be extended to behave like a sum type. This is much more verbose than a `choice`
+declaration, but gives the author full control over the representation and class
+members.
+
+The ability to create instances of the sum type can be straightforwardly
+emulated with factory functions and static constants (although the latter
+[are not yet designed](classes.md#no-static-variables)), and the internal
+storage layout will presumably involve untagged unions or some other low-level
+storage primitive which hasn't been designed yet, but the key to defining a sum
+type's interface is enabling it to support pattern matching. To do that, the sum
+type has to specify two things:
+
+-   The set of all possible alternatives, including their names and parameter
+    types, so that the compiler can typecheck the `match` body, identify any
+    unreachable `case`s, and determine whether any `case`s are missing.
+-   The algorithm that, given a value of the sum type, determines which
+    alternative is present, and specifies the values of its parameters.
+
+Here's how that would look if `Optional` were defined as a class:
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  fn None() -> Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some(value: T) -> ReturnType;
+    fn None() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
```

Comment:
The proposal used `Matchable`, not `Match`. Elsewhere we do things like `EqWith` so I don't think we should do this rename without a proposal.

Reply:
The PR descriptions points to this change coming from #1058

Reply:
`match` or `Matchable` is not mentioned on #1058.

AFAICT pointing to #1058 is a claim that Match is "Interfaces that contribute usefully to creating consistent APIs on types", leading to Matchable. I would argue it's instead "Interfaces that implement a specific language expression syntax" (i.e., `match`), leading to `Match` (the accepted proposal state, and similar to `ImplicitAs`) or `MatchWith`.

Reply:
Sorry, got that a little backwards: `Matchable` is the accepted proposal state, and would be the consistent API interface. `Match` is one argument to name under the second rule.

But I'm not trying to argue about how #1058 could be read here -- we have an accepted proposal.

Reply:
OK, I've promoted this to a proposal.

For clarity: is your concern purely procedural, or are you arguing that `Match.Op` is not the right name under #1058?

Reply:
My view is that `Match.Op` is a suitable name, consistent with the direction in #1058. I agree with @jonmeow that we should have this decision made in a proposal, so thanks for making this PR into one :)